### PR TITLE
Adv Tower lab config with included roles 

### PR DESCRIPTION
**Pull request by goetzrieger**
The main point of this PR is to include a number of roles into our config `ansible-automation-platform` as discussed with @tonykay and @rhjcd. This way we are free to adapt the roles local to our config and Tony can work on the "global" roles. These roles have been renamed and copied to a local `configs/ansible-automation-platform/roles/` folder:

- automation_controller_issue_cert
- tower-license-injector
- vscode-server
- deploy_automationcontroller
- tower_validate

We still use a number of default "global" roles from `ansible/roles`.

So while this PR is relatively large it should not touch on anything outside our config with one exception: `/.gitignore`. As we now need to **include** roles in our config dir, we put a "sub" `.gitignore` file into the `ansible-automation-platform` dir. For this to work we needed  to change ignoring `../roles`  to `../roles/*` so we could "un-ignore" in our config dir.

The ansible-automation-platform config can deploy a single-instance or a clustered Automation Controller env with vscode-server, LE certs and manifest injection.

### CODE REVIEW

File: ansible/configs/ansible-automation-platform/roles/automation_controller_issue_cert_cfg/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+          # tower may not be fully up yet, wait for it gratuitously
+          retries: 12
+          delay: 5
+      rescue:
+        - name: Try restarting tower
+          service:
+            name: automation-controller.service
+            state: restarted
+
+        - *tower-pinger-block
```
Comment (fridim) [inline ansible/configs/ansible-automation-platform/roles/automation_controller_issue_cert_cfg/tasks/main.yml:58]: i don't reckon any use of this "trick" elsewhere in agnosticd

I don't know if we have guidelines pro or against it. @tonykay  @jkupferer   FYI.

File: ansible/configs/ansible-automation-platform/roles/automation_controller_issue_cert_cfg/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    - name: make sure Ansible Tower is stopped
+      service:
+        name: automation-controller.service
+        state: stopped
+      register: stop_tower
+      until: stop_tower is not failed
+      retries: 5
+
+    # If this fails check out status of certbot: https://letsencrypt.status.io/
+    - name: ISSUE CERT
```
Comment (fridim) [inline ansible/configs/ansible-automation-platform/roles/automation_controller_issue_cert_cfg/tasks/main.yml:73]: Issue certificate*

File: ansible/configs/ansible-automation-platform/roles/automation_controller_issue_cert_cfg/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      shell: certbot certonly --no-bootstrap --standalone -d {{ ansible_host.split(".")[0] | lower }}.{{ subdomain_base }} --email ansible-network@redhat.com --noninteractive --agree-tos
```
Comment (fridim) [inline ansible/configs/ansible-automation-platform/roles/automation_controller_issue_cert_cfg/tasks/main.yml:74]: you're not using any feature of the `shell` module, `command` should be enough

Please wrap the line

```yaml
command: >-
  certbot certonly --no-bootstrap
  --standalone ...
```

File: ansible/configs/ansible-automation-platform/roles/deploy_automationcontroller_cfg/defaults/main.yml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+---
+
+deploy_automationcontroller_installer_url: "" # e.g. http://example.com/ansible-automation-setup-bundle-latest.tar.gz"
```
Comment (fridim) [inline ansible/configs/ansible-automation-platform/roles/deploy_automationcontroller_cfg/defaults/main.yml:3]: that default `""` won't work, right?
If it's the case, then i think it's preferred not to set it and make sure it's provided by the user using the role.

File: ansible/configs/ansible-automation-platform/roles/deploy_automationcontroller_cfg/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+# World's ugliest short term hack for fixing broken installer for clustering
+# dest: '/etc/systemd/system/automation-controller.service'
+
+# - name: Short term hack to fix Automation controller remote_src issue
+#   lineinfile:
+#     path: /tmp/automationcontroller_installer/collections/ansible_collections/ansible/automation_platform_installer/roles/automationcontroller/tasks/tasks.yml
+#     insertafter: ".*src: '/lib/systemd/system/automation-controller.service'"
+#     state: present
+#     line: "    remote_src: true"
```
Comment (fridim) [inline ansible/configs/ansible-automation-platform/roles/deploy_automationcontroller_cfg/tasks/main.yml:49]: can we remove those commented lines?

File: ansible/configs/ansible-automation-platform/roles/deploy_automationcontroller_cfg/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+  tags:
+    - deploy_automationcontroller_installer
+
+- name: Wait for Ansible controller to be up
+  uri:
+    url: "https://{{groups['automationcontroller'][0] }}/api/v2/ping/"
+    method: GET
+    user: "{{ deploy_automationcontroller_admin_user }}"
+    password: "{{ deploy_automationcontroller_admin_password }}"
+    validate_certs: false
```
Comment (fridim) [inline ansible/configs/ansible-automation-platform/roles/deploy_automationcontroller_cfg/tasks/main.yml:72]: any easy way to remove `validate_certs: false` here?  like adding the cert using update-ca-trust or similar
or maybe just the `ca_path` of the uri module.  If easy, if not, it's fine.

File: ansible/configs/ansible-automation-platform/roles/deploy_automationcontroller_cfg/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+#       uri:
+#         url: "https://{{groups['automationcontroller'][0] }}/api/v2/config/"
+#         method: POST
+#         user: "{{ deploy_automationcontroller_admin_user | default('admin') }}"
+#         password: "{{ deploy_automationcontroller_admin_password }}"
+#         body: '{ "eula_accepted": true, "manifest": "{{ r_automationcontroller_manifest_file.content }}" }'
+#         body_format: json
+#         validate_certs: false
+#         force_basic_auth: true
+
```
Comment (fridim) [inline ansible/configs/ansible-automation-platform/roles/deploy_automationcontroller_cfg/tasks/main.yml:106]: remove comments if not used

File: ansible/configs/ansible-automation-platform/roles/tower-license-injector_cfg/tasks/manifest.yml [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+    file: "man_{{ tower_license_manifest.type }}.yml"
+
+- name: manifest | inject tower license manifest
+  awx.awx.tower_license:
+    manifest: "{{ __tower_license_path }}"
+    #eula_accepted: true  # works only for Tower 3.8 / awx.awx 19.1.0
+    tower_host: "https://localhost/"
+    tower_username: admin
+    tower_password: "{{ hostvars[groups['bastions'][0]]['tower_admin_password'] }}"
+    tower_verify_ssl: false
```
Comment (fridim) [inline ansible/configs/ansible-automation-platform/roles/tower-license-injector_cfg/tasks/manifest.yml:14]: same question, can we easily remove `tower_verify_ssl`?

File: ansible/configs/ansible-automation-platform/roles/tower-license-injector_cfg/templates/tower_cli.j2 [hunk 9]
```diff
@@ -0,0 +1,5 @@
+[general]
+host = "{{ tower_hostname }}"
+username = admin
+password = "{{ hostvars[groups['bastions'][0]]['tower_admin_password'] }}"
+verify_ssl = False
```
Comment (fridim) [inline ansible/configs/ansible-automation-platform/roles/tower-license-injector_cfg/templates/tower_cli.j2:5]: same question for `verify_ssl`

File: ansible/configs/ansible-automation-platform/roles/tower_validate_cfg/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,66 @@
+---
+# tasks file for tower_validate
+- name: ping the Tower using the API
+  uri:
+    url: "https://{{ tower_hostname }}/api/v2/ping/"
+    validate_certs: "{{ tower_validate_certs }}"
+    method: GET
+    headers:
+      #Content-Type: "application/json"
+      Authorization: "Basic {{ (tower_admin_user + ':' + tower_admin_password) | b64encode }}"
```
Comment (fridim) [inline ansible/configs/ansible-automation-platform/roles/tower_validate_cfg/tasks/main.yml:10]: do we have to craft the headers? can't we use basic auth from the `uri` module directly ?

File: ansible/configs/ansible-automation-platform/roles/vscode-server_cfg/tasks/nginx-server.yml [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        disable_gpg_check: true
+
+    - name: install nginx/certbot
+      yum:
+        name:
+          - nginx
+          - certbot
+          - python3-certbot-nginx
+        state: present
+  when: ansible_distribution == "RedHat" and ansible_distribution_major_version == '8'
```
Comment (fridim) [inline ansible/configs/ansible-automation-platform/roles/vscode-server_cfg/tasks/nginx-server.yml:27]: if you fix the GPG check issue beforehand, you should be able to just use the `package` ansible module and ansible will pick the default package manager, right?

**Review by fridim — CHANGES_REQUESTED** [review]

**Comment by fridim** [comment]
Also todo, remove those trailing spaces:

```
/home/runner/work/agnosticd/agnosticd/ansible/configs/ansible-automation-platform/env_vars.yml
  50:1      error    trailing spaces  (trailing-spaces)
  54:24     error    trailing spaces  (trailing-spaces)
  55:1      error    trailing spaces  (trailing-spaces)
```

File: ansible/configs/ansible-automation-platform/post_infra.yml [hunk 12]
```diff
@@ -13,7 +13,7 @@
       url: "https://{{ ansible_tower_ip }}/api/v1/job_templates/{{ job_template_id }}/launch/"
       method: POST
       user: "{{tower_admin}}"
-      password: "{{tower_admin_password}}"
+      password: "{{ hostvars['bastion[0]']['ansible_facts'][tower_admin_password] }}"
```
[the remark this answers is not in the document]
Comment (goetzrieger) [reply ansible/configs/ansible-automation-platform/post_infra.yml:16]: TBH no idea what this task is doing in the first place, it was already there in post_infra.yml and should not be called anyway. `"{{ hostvars['bastion[0]'][tower_admin_password] }}"` would be right to get the pw of the Tower we are about to install, it made it here through find/replace I'm afraid. ;)

[the remark this answers is not in the document]
Comment (goetzrieger) [reply ansible/configs/ansible-automation-platform/post_infra.yml:16]: Can we remove this task or is it needed somehow by agnosticd?

File: ansible/configs/ansible-automation-platform/post_software.yml [hunk 13]
```diff
@@ -7,39 +7,39 @@
...
+#       vars:
+#         git_user: git
+#         git_project: structured-content
+#         git_authorized_keys: "{{ output_dir }}/{{ guid }}_{{ ansible_user }}_id_rsa.pub"
+#   post_tasks:
+#     - name: add user {{ ansible_user }} to {{ git_user }} group
+#       user:
+#         name: "{{ ansible_user }}"
+#         groups: "{{ git_user }}"  # it has been created by the git-server-ssh role
+#         append: true
```
[the remark this answers is not in the document]
Comment (goetzrieger) [reply ansible/configs/ansible-automation-platform/post_software.yml:42]: Sure

File: ansible/configs/ansible-automation-platform/roles/automation_controller_issue_cert_cfg/tasks/main.yml (hunk 1 above)
Comment (goetzrieger) [reply ansible/configs/ansible-automation-platform/roles/automation_controller_issue_cert_cfg/tasks/main.yml:58]: Role comes straight from the Ansible Workshop Provisioner. Works and we would prefer not having to develop new stuff.

File: ansible/configs/ansible-automation-platform/roles/deploy_automationcontroller_cfg/tasks/main.yml (hunk 7 above)
Comment (goetzrieger) [reply ansible/configs/ansible-automation-platform/roles/deploy_automationcontroller_cfg/tasks/main.yml:106]: Sure

**Comment by fridim** [comment]
i just noticed you're touching `.gitignore` and that you have a role in the config `roles` directory.

as specified here: https://github.com/redhat-cop/agnosticd/blob/development/docs/How_to_use_galaxy_roles_in_your_config.adoc#development-workflow

The `roles` directory in configs is used for dynamic roles installed via collections or galaxy. We don't use them to store roles for a config, hence the line in .gitignore.

Can you move your roles in the `ansible/roles` directory? You can also import them from external repositories using the file `requirements.txt` in the config dir.

File: .gitignore [hunk 14]
```diff
@@ -15,7 +15,7 @@
 ansible/workdir/*
 ansible/judd.sh
 ansible/collections
-ansible/configs/*/roles
+ansible/configs/*/roles/*
```
Comment (fridim) [inline .gitignore:18]: can you explain why this change is needed?

see https://github.com/redhat-cop/agnosticd/blob/development/docs/How_to_use_galaxy_roles_in_your_config.adoc#development-workflow

Comment (fridim) [reply .gitignore:18]: OK  sorry, just saw it in the description

> So while this PR is relatively large it should not touch on anything outside our config with one exception: /.gitignore. As we now need to include roles in our config dir, we put a "sub" .gitignore file into the ansible-automation-platform dir. For this to work we needed to change ignoring ../roles to ../roles/* so we could "un-ignore" in our config dir.

Let me discuss it internally

Comment (fridim) [reply .gitignore:18]: after discussing with @wkulhanek  and @tonykay, we don't think it's a good idea to allow this practice. Documentation about `ansible/configs/*/roles` is pretty clear, it's reserved for dynamic roles imported by ansible galaxy or collections.

You have the choice to:
* move the roles to ansible/roles 
* externalize the roles in another repository and import them using the requirements.txt file in the config directory

File: ansible/configs/ansible-automation-platform/roles/automation_controller_issue_cert_cfg/tasks/main.yml (hunk 1 above)
Comment (fridim) [reply ansible/configs/ansible-automation-platform/roles/automation_controller_issue_cert_cfg/tasks/main.yml:58]: as disccussed, that role is temporary for summit and this is taken from ansible workshop

**Comment by ericzolf** [comment]
Fix agreed changes with Guillaume

- move _cfg roles to standard roles directory (adding warning banner to README.md/adoc)
- rolled back unnecessary changes to .gitignore and post_infra.yml
- removed one trailing blank in sample_vars.yml
- use more standard approach to uri basic auth

and then two follow-up fixes:
1. Add force_basic_auth to tower_validate{_cfg} role to avoid Bearer not being supported error
2. Fix line too long complain from ansible-lint

File: .gitignore (hunk 14 above)
Comment (ericzolf) [reply .gitignore]: It is rolled back (just the ending line is still there)

File: ansible/configs/ansible-automation-platform/roles/tower_validate_cfg/tasks/main.yml (hunk 10 above)
Comment (ericzolf) [reply ansible/configs/ansible-automation-platform/roles/tower_validate_cfg/tasks/main.yml]: Fixed with basic auth and enforcing basic auth.

File: ansible/configs/ansible-automation-platform/post_infra.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (ericzolf) [reply ansible/configs/ansible-automation-platform/post_infra.yml]: I just rolled back the change.

**Review by fridim — APPROVED** [review]
